Rework overview/what are extensions page (#411)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/mv3/overview/index.md

description: An overview of the purpose of Chrome Extensions and how they're developed.
---

This page provides a very brief introduction to what extensions are and how you can make a trivial one.

## About extensions

Extensions are small software programs that customize the browsing experience. They let users tailor
Chrome functionality and behavior in many ways, providing things like:

* Productivity tools
* Web page content enrichment
* Information aggregation
* Fun and games

These are just a few examples of the many things that extensions can do. See the [Chrome Web
Store][5] to see thousands of different examples of published extensions.

### How do extensions work?

Extensions are built on web technologies such as HTML, JavaScript, and CSS.  They run in a separate,
sandboxed execution environment and interact with the Chrome browser.  Consider the context for this
-- Chrome runs on a device, connecting the user to resources on the Web, as depicted in this context
diagram:

> {% Img src="image/SHhb2PDKzXTggPGAYpv8JgR81pX2/5qVlqGn6vVMCWCSRclOY.png", alt="Context diagram of Chrome with an extension", width="517", height="476" %}

Reviewer: I'm not sure I understand this diagram.  Is the extension a part of Chrome?  Does it not have access to HTML APIs, storage, and UI (which are separate)?  Can it not access the internet?

Can you describe what the main takeaway we want the reader to have from this diagram is?

Author: What I'd like to depict is that Chrome does its thing as a browser, and then extensions "plug onto" Chrome and have access (via the API) of selected Chrome features to intercede in the way they work.

Author: For things like this I miss whiteboarding in the office...

Reviewer: Hmm... Honestly, for now, I wonder if it's better to just omit this.  I don't think, in its current form, the diagram adds much information.  I wonder if just describing this (similar to what you have) would be sufficient?

Extensions are built on web technologies such as HTML, JavaScript, and CSS.  They run in a separate,
sandboxed execution environment and interact with the Chrome browser.  Extensions are designed to "extend" the Chrome browser -- they utilize APIs in order to modify browser behavior and access web content.  There are two aspects to consider:

WDYT?

Author: Yeah, I wanted a diagram to a) help break up the slabs of text and b) help construct context. But I am fine with suppressing this for now and trying to come up with something more meaningful later.

Also shown in the diagram is an extension "plugged into" the Chrome browser. The extension can
access and modify selected resources and actions in Chrome. There are two aspects to consider:

> **The extensions mechanism**&emsp;This provides a consistent way for users to install and activate

Reviewer: I don't love the phrasing of this as "the extensions mechanism" - it's not clear to me what this is referring to, or what the bounds of it are.  If this is just the installation and activation, it's more the UI than the "mechanism".  Can you add a bit more detail here about what this is trying to get at?

Author: Done: Changed it to "extensions user interface" and made it about "manage their extensions"

extensions.

**Extensions APIs**&emsp;The [extensions APIs](/docs/extensions/reference/) allow the extension's

> code to access features of the browser itself, activating tabs, modifying net requests, and so on.

Reviewer: ```suggestion
code to access features of the browser itself: activating tabs, modifying net requests, and so on.
```

(or an em dash, or insert a "such as")

> To create an extension, you assemble some resources -- JavaScript and HTML files, a manifest,

Reviewer: nit: maybe lead with "manifest", since it's the only required bit?

Author: Done

images, and others -- that constitute the extension. For development and testing, you can load these
"unpacked" into Chrome using extension developer mode. Once you are happy with your extension, you
can package it and distribute it to users.

### How do users get extensions?

Most Chrome users get extensions from the Chrome Web Store (CWS). Developers across the globe
publish their extensions in the Chrome Web Store where they are reviewed and made available to end
users.

{% Aside %}

Some organizations use enterprise policies to install extensions on their user's devices. These
extensions may either be fetched from the Chrome Web Store or hosted on the organization's web servers.
{# TODO: See XXX for info on the capabilities and limitations of self hosting. #}

{% endAside %}

You can distribute your extensions through the [Chrome Developer Dashboard][4], publishing them to the [Chrome
Web Store][5]. For more information, see the Chrome Web Store [developer documentation][6].

### A note about extensions policy

> Extensions on the Chrome Web Store must adhere to the [Chrome Web Store policies][cws-policies].

Reviewer: nit: Can we use a consistent link style?  (I'd lean towards [named-style] instead of [numbered-style], since it's much more clear IMO)

Reviewer 2: The main reason we currently use numbers is that our automated tooling assigned numbers when we ported content over. I don't have any objection to normalizing on named-style. If we do that I'd favor sorting links in the footer alphabetically.

Reviewer: Yep, and I don't think we need to go on a spree updating each individual page.  But if we're heavily modifying this one already, I think we should fix it up while we're here. : )

Author: And I often like [inline-style](...) ;-)

I have updated the links in this doc to be all [named-style] and alphabetized the definitions part.

Here are some things to keep in mind as you begin:

* An extension must fulfill a [single purpose][1] that is narrowly defined and easy to understand. A
single extension can include multiple components and a range of functionality, as long as everything
contributes towards a common purpose.

{% Img src="image/SHhb2PDKzXTggPGAYpv8JgR81pX2/XniXB3snAeMvLwI1am3O.png", alt="Screenshot of AMP validator extension pinned", width="169", height="62" %}

* User interfaces should be minimal and have intent. They can range from a simple icon, such as the
  [AMP validator] extension shown above, to opening a new window with a form, like the [PackageTrack]

>   extension shown below.

Reviewer: Can we use a google-published extension, rather than PackageTrack?

Reviewer 2: +1. If we do want to use 3rd party extensions we should probably discuss offline what steps are required to do so.

Author: @dotproto Any particular recommendation?

Author: Replaced with Google Similar Pages. I don't like the "preview unavailable" images, so I'm still shopping...

> description: An overview of the purpose of Chrome Extensions and how they're developed.

Reviewer: (random comment spot)

On the netlify preview, the link from https://deploy-preview-411--developer-chrome-com.netlify.app/docs/extensions/mv3/ to "what are extensions" is broken, but it looks fine in production.  Is that an issue with this PR, or with netlify?  Or where the link is technically wrong, but the production server catches and redirects?  (That's my closest guess?)

I can't see how this PR would have broken it, since it's not moving the file, but figured I'd raise it.

Author: Yeah, the link is wrong and redirecting in prod and on localhost but not in the preview. (Ran into this before where netlify doesn't do our redirects.) Fixed the links for good form.

---

> This page provides a very brief introduction to what extensions are and how you can make a trivial one.

Author: This kind of intro is standard tradecraft and eases the user into the page and its context. Doesn't matter that you could figure it out. Also appropriate that it's similar to, and lightly expands on, the description line: these appear in different places entirely. I do like the rewording for this bit of body text.

-- Chrome runs on a device, connecting the user to resources on the Web, as depicted in this context
diagram:

{% Img src="image/SHhb2PDKzXTggPGAYpv8JgR81pX2/5qVlqGn6vVMCWCSRclOY.png", alt="Context diagram of Chrome with an extension", width="517", height="476" %}

Also shown in the diagram is an extension "plugged into" the Chrome browser. The extension can
access and modify selected resources and actions in Chrome. There are two aspects to consider:

**The extensions user interface**&emsp;This provides a consistent way for users to manage their
extensions.

**Extensions APIs**&emsp;The [extensions APIs](/docs/extensions/reference/) allow the extension's
code to access features of the browser itself: activating tabs, modifying net requests, and so on.

To create an extension, you assemble some resources -- a manifest, JavaScript and HTML files,
images, and others -- that constitute the extension. For development and testing, you can load these
"unpacked" into Chrome using extension developer mode. Once you are happy with your extension, you
can package it and distribute it to users.

### How do users get extensions?

> Most Chrome users get extensions from the Chrome Web Store (CWS). Developers across the globe

Reviewer: ```suggestion
Most Chrome users get extensions from the Chrome Web Store. Developers across the globe
```

We don't use CWS anywhere else in this article, or generally in the docs AFAIK.  It's not really a popular abbreviation outside of Google, I think.  (Doing a google search for "CWS" or even "CWS extension" doesn't really show much)

Author: Done.

* Productivity tools
* Web page content enrichment
* Information aggregation
* Fun and games

These are just a few examples of the many things that extensions can do. See the [Chrome Web
Store][cws] to see thousands of different examples of published extensions.

### How do extensions work? {: #basics }

Extensions are built on web technologies such as HTML, JavaScript, and CSS.  They run in a separate,
sandboxed execution environment and interact with the Chrome browser.

{# TODO: put a diagram here that is useful and helps with context #}
{% if false %}

Consider the context for this
-- Chrome runs on a device, connecting the user to resources on the Web, as depicted in this context
diagram:

{% Img src="image/SHhb2PDKzXTggPGAYpv8JgR81pX2/5qVlqGn6vVMCWCSRclOY.png", alt="Context diagram of Chrome with an extension", width="517", height="476" %}

Also shown in the diagram is an extension "plugged into" the Chrome browser. The extension can
access and modify selected resources and actions in Chrome. There are two aspects to consider:
{% endif %}

Extensions let you "extend" the browser by using APIs to modify browser behavior and access web

> content. There are two aspects to consider:

Reviewer 2: > There are two aspects to consider:

With regard to what?

Author: Rephrased this in 23fcebf

**The extensions user interface**&emsp;This provides a consistent way for users to manage their

> extensions.

Reviewer 2: I'm a little confused by this section. On one hand I feel like it's trying to describe elements that the extension developer needs to consider, on the other the actual text seems to be about browser level considerations.

What's the objective of this section? Asking in order to see if I have any ideas on how we might be able to rephrase this to better address that goal.

Author: This is part of understanding what extensions are -- knowing the user experience is part of that. I'd like to get this merged, so can we steer away from optimization for now?

Author: I would prefer to gather further improvements in a follow-up ticket.

Reviewer 2: That seems reasonable. I won't let good be the enemy of perfect here 😄

**Extensions APIs**&emsp;The [extensions APIs](/docs/extensions/reference/) allow the extension's
code to access features of the browser itself: activating tabs, modifying net requests, and so on.

To create an extension, you assemble some resources -- a manifest, JavaScript and HTML files,
images, and others -- that constitute the extension. For development and testing, you can load these

> "unpacked" into Chrome using extension developer mode. Once you are happy with your extension, you

Reviewer 2: I think "developer mode" link to a related page. Perhaps https://developer.chrome.com/docs/extensions/mv2/tut_debugging/#debug_bg ?

Author: Done in 23fcebf

> can package it and distribute it to users.

Reviewer 2: Perhaps we should link to packaging and distribution information, maybe https://developer.chrome.com/docs/webstore/publish/ ?

Author: Done in 23fcebf

### How do users get extensions? {: #getting-extensions }

> Most Chrome users get extensions from the Chrome Web Store. Developers across the globe

Reviewer 2: Make "Chrome Web Store" link to http://chrome.google.com/webstore

Author: Done in 23fcebf

publish their extensions in the Chrome Web Store where they are reviewed and made available to end
users.

{% Aside %}

> Some organizations use enterprise policies to install extensions on their user's devices. These

Reviewer 2: Make "enterprise policies" link to https://cloud.google.com/docs/chrome-enterprise/policies/

Author: Done in 23fcebf

description: An overview of the purpose of Chrome Extensions and how they're developed.
---

This page provides a very brief introduction to what extensions are and how you can make a trivial one.

## About extensions

Extensions are small software programs that customize the browsing experience. They let users
tailor Chrome functionality and behavior in many ways, providing things like:

> * Games

Reviewer 2: I'm very hesitant about having "games" be the first example of what extensions can do. Honestly I'm not even sure if we should include it in this list. 

Broadly speaking extensions are meant as a way to extend your web browser's functionality. You might want to customize how a feature works (e.g. replace the browser's bookmark manager), add new functionality (e.g. look up a word by highlighting it), integrate tools that keep you on track (e.g. blocking distracting websites), and so on.

While there are some games in the Chrome Web Store, I don't think this is the right focus for the extensions platform. Generally speaking I think games are better as web sites rather tightly integrated into the browser experience.

Reviewer 3: Could we say "Delightful experiences" aka something like "Fun", instead of "Games"?

This would describe to me [this extension](https://chrome.google.com/webstore/detail/tabby-cat/mefhakmgclhhfbdadeojlkbllmecialg?hl=en).

Author: Changed this to "Fun & games" as the last bullet.

> User interfaces should be minimal and have intent. They can range from a simple icon, such as the [Google
> Mail Checker extension][2] shown on the right, to [overriding][3] an entire page.

Reviewer 2: ```suggestion
User interfaces should be minimal and have intent. They can range from a simple icon, such as the [Google
Mail Checker extension][2] shown above, to [overriding][3] an entire page.
```

Also, can we update the screenshot of the Google Mail checker image? Chrome's UI has changed since this image was captured.

### How do users get extensions?

When you make an extension, you zip it into a single `.crx` package that the user downloads and installs. This

> means extensions do not depend on content from the web, unlike ordinary web apps.

Reviewer 2: I suspect that the original version used the passive voice because CWS is typically the entity that generates the crx package that a user downloads.

I think it may be better to pivot this section (+1 to creating a new heading) a little to focus more on CWS. While we may want to talk about what a CRX file is, I'm not sure it fits under this heading. To that end here's a quick attempt to draft up a new body for this section.

> Most Chrome users get extensions from the Chrome Web Store (CWS). Developers across the globe submit their extension to the Chrome Web Store where they are reviewed and made available to end users. 
> 
> Some businesses and schools use [enterprise policies](https://cloud.google.com/docs/chrome-enterprise/policies) to install extensions on their user's devices. These extensions are either fetched from the Chrome Web Store or they may be hosted on web servers. See XXX for info on the capabilities and limitations of  self hosting.

Author: Adapted the suggested text to replace the content of this section.

3.  Keep up to date by reading the [Chromium blog][12]
4.  Subscribe to the [chromium-extensions group][13]

## Featured videos {: #Featured-videos }

[Technical videos][14]

> [Developer snapshots][15]

Reviewer 2: +1 to dropping this.

Much broader topic, but later in the year (Q2 maybe?) we should consider what video content might be worth producing.

> ## What next? {: #How-do-I-start }

Author: Acknowledged.

## What next? {: #How-do-I-start }

1.  Follow the [Getting Started tutorial][10]
1.  Explore the [extension samples][2]

> 1.  Subscribe to the [chromium-extensions group][13]

Reviewer: ```suggestion
2.  Subscribe to the [chromium-extensions group][13]
```

Author: It's better Markdown practice to make numbered lists go 1., 1., 1., ... and let the processor handle the numbering. Check the staging preview.

Reviewer: Nifty!  I saw that the rendering was correct, but didn't realize this was a feature of markdown.  Good to know.
